File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.cpp
#include <vector>

using namespace std;

long long max_sum(vector<int>& arr){
    vector<long long> sum(arr.size(), 0ll);
    sum[0] = arr[0];
    long long result = arr[0];
    
    for(int i = 1; i < arr.size(); i++){
        sum[i] = max(0ll, sum[i - 1]) + arr[i];
        result = max(result, sum[i]);
    }
    
    return result;
}

long long solution(vector<int> sequence) {
    vector<int> arr1(sequence);
    for(int i = 0; i < arr1.size(); i+= 2) arr1[i] *= -1;
    
    vector<int> arr2(sequence);
    for(int i = 1; i < arr2.size(); i+= 2) arr2[i] *= -1;
    
    return max(max_sum(arr1), max_sum(arr2));
}

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.py
import copy

def max_sum(arr):
    now_max = [arr[0]]
    result = arr[0]
    for i in range(1, len(arr)):
        now_max.append(max(0, now_max[i - 1]) + arr[i])
        result = max(result, now_max[i])
    
    return result

def solution(sequence):
    seq1 = copy.deepcopy(sequence)
    seq2 = copy.deepcopy(sequence)
    
    for i in range(len(sequence)):
        if i % 2 == 0:
            seq1[i] *= -1
        else:
            seq2[i] *= -1
    
    return max(max_sum(seq1), max_sum(seq2))
